pkg/endpoint/regenerationcontext.go

```diff
@@ -0,0 +1,34 @@
+// Copyright 2016-2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package endpoint
+
+// ExternalRegenerationMetadata contains any information about a regeneration that
+// the endpoint subsystem should be made aware of for a given endpoint.
+type ExternalRegenerationMetadata struct {
```

Comment:
Yes, an example might let me prove to myself why what I'm saying makes no sense.
Right now, our code does something like:
- Make a `ExternalRegenerationMetadata` with a reason (e.g. in `daemon.TriggerReloadWithoutCompile`, `daemon.TriggerPolicyUpdate` and `daemon.regenerateRestoredEndpoints`)
- Call `endpointmanager.RegenerateAllEndpoints` or `Endpoint.Regenerate` directly. We end up in `endpoint.Regenerate` anyway.
- `.Regenerate` initialises a `regenerationContext` from the `ExternalRegenerationMetadata`
- various fields are used and set in `regenerationContext` and `datapathRegenerationContext`. E.g. DoneFunc in `Endpoint.Regenerate` [1]

My proposal (including the other one about constructors) was to:
- External caller calls `newRegenerationContext(reason string, reloadDatapath bool)`. This also allocates all fields internally, including `datapathRegenerationContext`.
- Call `endpointmanager.RegenerateAllEndpoints` or `Endpoint.Regenerate` directly. We end up in `endpoint.Regenerate` anyway.
- `.Regenerate` doesn't need to initialise any fields, but can set things like `DoneFunc`

The idea here was that the caller of .Regenerate needed to provide a context, since it was using that package. The package gives it a type that works for this purpose (this is also true of `ExternalRegenerationMetadata`). This single type is ready to go as soon as it is passed to `pkg/endpoint`'s functions.
As you mentioned offline, we initialise things like timeouts and `context.Context` and those are more sensitive to where they are created. This generally makes what I'm saying less appealing, since we would be inconsistent in how some fields are set/created. I did notice that  `datapathRegenerationContext` is what holds these time-sensitive fields, and so it could be created later in the process. This is basically `prepareForProxyUpdates` and so it's acting as a constructor of sorts for this (many fields are set differently).
Anyway, in writing this up I realised how we ended up with these types. I still think we can get away with only `regenerationContext` (but package public) and `datapathRegenerationContext` but this is about 10 lines different from this PR, so not so worth it.


[1] I just noticed that this isn't used outside pkg/endpoint. Can it be private?

Reply:
The reason I wanted to make two separate types, one external and internal, is that one can be shared by all Endpoints (`ExternalRegenerationMetadata`) vs. one that is local to a single goroutine for a regeneration for a given endpoint `regenerationContext`.  If i follow your proposal, the external caller would allocate the `datapathRegenerationContext`, which I don't want, because I don't want the external caller to be able to instantiate that field because it should only be managed by the goroutine that is regenerating the endpoint.